### PR TITLE
chore: finish enabling no-unnecessary-condition internally

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## PR Checklist

- [x] Addresses an existing open issue: continues #7338
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Follows #7837 et al. in fixing up existing [@typescript-eslint/no-unnecessary-condition](https://typescript-eslint.io/rules/no-unnecessary-condition) violations. Also enables the lint rule overall.

### CODE REVIEW

File: packages/eslint-plugin/src/rules/consistent-type-exports.ts [hunk 1]
```diff
@@ -109,6 +109,7 @@ export default createRule<Options, MessageIds>({
       ExportNamedDeclaration(node: TSESTree.ExportNamedDeclaration): void {
         // Coerce the source into a string for use as a lookup entry.
         const source = getSourceFromExport(node) ?? 'undefined';
+        // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/consistent-type-exports.ts:112]: This is the old issue we've got filed right where TS is giving the wrong type for logical assignment operators?

If yes - Oof we should prioritise this #6762

File: packages/eslint-plugin/src/rules/prefer-readonly.ts [hunk 2]
```diff
@@ -105,7 +105,7 @@ export default createRule<Options, MessageIds>({
     function isDestructuringAssignment(
       node: ts.PropertyAccessExpression,
     ): boolean {
-      let current: ts.Node = node.parent;
+      let current = node.parent as ts.Node | undefined;
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/prefer-readonly.ts:108]: i wonder if we should patch the TS types here?

File: packages/website/src/components/Playground.tsx [hunk 3]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable @typescript-eslint/no-unnecessary-condition */
```
Comment (bradzacher) [inline packages/website/src/components/Playground.tsx:1]: ditto were there a lot here?

File: packages/website/src/components/typeDetails/TypesDetails.tsx [hunk 4]
```diff
@@ -51,17 +51,15 @@ export function TypesDetails({
         </div>
       </Panel>
       <PanelResizeHandle className={styles.PanelResizeHandle} />
-      {selectedNode && (
```
Comment (bradzacher) [inline packages/website/src/components/typeDetails/TypesDetails.tsx:54]: i *vaguely* remember armano saying this was a necessary condition?
might be worth double checking the preview deploy

File: packages/eslint-plugin/src/rules/consistent-type-exports.ts (hunk 1 above)
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/src/rules/consistent-type-exports.ts:112]: Yeah :/

File: packages/eslint-plugin/src/rules/prefer-readonly.ts (hunk 2 above)
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/src/rules/prefer-readonly.ts:108]: Yeah that would be nice. I've played around with trying to make parent optional once in a while but it's pretty much blocked on https://github.com/microsoft/TypeScript/issues/56275. I'll note that here.

File: packages/rule-tester/src/RuleTester.ts [hunk 5]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable @typescript-eslint/no-unnecessary-condition */
```
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply packages/rule-tester/src/RuleTester.ts:1]: Yeah, 9 of them. Mostly around type safety checks on user-provided parameters. E.g. `config == null`:

```ts
    if (typeof config !== 'object' || config == null) {
      throw new TypeError(
        'RuleTester.setDefaultConfig: config must be an object',
      );
    }
```

File: packages/website/src/components/Playground.tsx (hunk 3 above)
Comment (JoshuaKGoldberg) [reply packages/website/src/components/Playground.tsx:1]: Yeah. They should be improved by https://github.com/typescript-eslint/typescript-eslint/issues/7630. I'll mark that here.

File: packages/website/src/components/typeDetails/TypesDetails.tsx (hunk 4 above)
Comment (JoshuaKGoldberg) [reply packages/website/src/components/typeDetails/TypesDetails.tsx:54]: Works locally and on the preview for me 🚀

**Review by bradzacher — APPROVED** [review]

**Pull request merged**